### PR TITLE
feat(planner): support position function

### LINKED ISSUE #5536: Feature: Support `POSITION` expression in new planner
The reference https://databend.rs/doc/reference/functions/string-functions/position

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

as title

## Changelog

- New Feature

## Related Issues

Fixes #5536

### CODE REVIEW

File: query/src/sql/planner/semantic/type_check.rs [hunk 1]
```diff
@@ -899,4 +905,47 @@ impl<'a> TypeChecker<'a> {
...
+                value: index.clone(),
+                data_type: index.data_type(),
+            }
+            .into(),
+            index.data_type(),
+        ))
+    }
+
+    fn match_expr(&self, expr: &'_ Expr<'a>) -> Result<String> {
+        let expr_str = match expr {
```
Comment (xudong963) [inline query/src/sql/planner/semantic/type_check.rs:931]: For the sql: `SELECT POSITION("xbar" IN "xbar0");`
`"xbar"` will be identified as `Expr::ColumnRef`, is this the correct result for parser? @andylokandy  If so, I need to handle the `ColumnRef` case here.

File: query/src/sql/planner/semantic/type_check.rs [hunk 2]
```diff
@@ -899,4 +905,47 @@ impl<'a> TypeChecker<'a> {
             data_type,
         ))
     }
+
+    async fn resolve_position(
+        &self,
+        substr_expr: &'_ Expr<'a>,
+        str_expr: &'_ Expr<'a>,
```
Comment (leiysky) [inline query/src/sql/planner/semantic/type_check.rs:912]: Maybe it's possible to remove the lifetime annotation `'_`?

File: query/src/sql/planner/semantic/type_check.rs [hunk 3]
```diff
@@ -899,4 +905,47 @@ impl<'a> TypeChecker<'a> {
...
+            .into(),
+            index.data_type(),
+        ))
+    }
+
+    fn match_expr(&self, expr: &'_ Expr<'a>) -> Result<String> {
+        let expr_str = match expr {
+            Expr::Literal { lit, .. } => {
+                let (value, data_type) = self.resolve_literal(lit, None)?;
+                if !data_type.eq(&DataTypeImpl::String(StringType {})) {
```
Comment (leiysky) [inline query/src/sql/planner/semantic/type_check.rs:934]:
```suggestion
                if data_type != &StringType::new_impl() {
```

File: tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql [hunk 4]
```diff
@@ -240,4 +240,9 @@ select '====Correlated Subquery====';
 select * from numbers(10) as t where exists (select * from numbers(2) as t1 where t.number = t1.number);
 select (select number from numbers(10) as t1 where t.number = t1.number) from numbers(10) as t order by number;
 
+-- position function
+select '===Position Function===';
+SELECT POSITION('bar' IN 'foobarbar');
```
Comment (leiysky) [inline tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:245]: Shall we allow it to take variable as argument? Like `position('bar' in column)`

Comment (xudong963) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:245]: I checked the old planner, it supports the case.

File: query/src/sql/planner/semantic/type_check.rs (hunk 1 above)
Comment (leiysky) [reply query/src/sql/planner/semantic/type_check.rs:931]: You can replace double quote `"` with single quote `'`.

File: tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql (hunk 4 above)
Comment (andylokandy) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:245]: This works in pg:

```SQL
drop table if EXISTS T;
create table t (a varchar, B varchar);
insert into t values ('FOO', 'barfoo');
select position(UPPER(A) in UPPER(B)) from T;
```

**Review by sundy-li — APPROVED** [review]

**Review by andylokandy — APPROVED** [review]

**Pull request merged**